File: algorithms/2_medium/R/01190_reverse_substrings_between_each_pair_of_parentheses/stack.c
typedef ptrdiff_t isize;

static void reverse(char* s, isize lo, isize hi)
{
    while (lo < hi) {
        const char temp = s[lo];
        s[lo] = s[hi];
        s[hi] = temp;
        lo += 1;
        hi -= 1;
    }
}

char* reverseParentheses(char* s)
{
    assert(s != NULL && s[0] != '\0');
    const isize len = (isize)strlen(s);
    assert(len <= 2000);
    int16_t* stack = malloc((size_t)len * sizeof(*stack));
    char* res = malloc((size_t)(len + 1) * sizeof(*res));
    assert(res != NULL && stack != NULL);
    isize res_len = 0;
    isize stack_len = 0;

    for (isize i = 0; i < len; i += 1) {
        if (s[i] == '(') {
            stack[stack_len] = (int16_t)res_len;
            stack_len += 1;
        } else if (s[i] == ')') {
            assert(stack_len > 0 && "invalid input");
            isize lo = stack[stack_len - 1];
            stack_len -= 1;
            isize hi = res_len - 1;
            reverse(res, lo, hi);
        } else {
            res[res_len] = s[i];
            res_len += 1;
        }
    }

    res[res_len] = '\0';

    return res;
}

File: algorithms/2_medium/R/01190_reverse_substrings_between_each_pair_of_parentheses/stack.rs
impl Solution {
    pub fn reverse_parentheses(s: String) -> String {
        let bytes = s.as_bytes();
        let len = bytes.len();
        assert!(len > 0 && len <= 2000);
        let mut stack: Vec<u16> = Vec::with_capacity(len);
        let mut res: Vec<u8> = Vec::with_capacity(len);

        for &b in bytes {
            match b {
                b'(' => {
                    stack.push(res.len() as u16);
                },
                b')' => {
                    let lo = stack.pop().expect("invalid input") as usize;
                    res[lo..].reverse();
                },
                _ => {
                    res.push(b);
                }
            }
        }

        String::from_utf8(res).unwrap()
    }
}
